### DISCUSSION TITLE
Many to many relationship : Have multiple times the same relation

### DISCUSSION THREAD

**Discussion by wazazaby** [prisma/prisma#5922] [Q&A]
Hi!

I'm currently writing an app for school with Node / Fastify and Prisma.

I have an implicit many-to-many relationship between two models (they're called Component and Module). A module is composed of multiple components, and can be linked to the same component multiple times, and I don't know how do to this... I've searched through SO and the documentation but I haven't found anything 😢 

Here are my models : 
```prisma
model Module {
    id                  Int                 @default(autoincrement()) @id
    label               String
    reference           String              @unique
    shortDescription    String?             @db.Text
    description         String              @db.MediumText
    createdAt           DateTime            @default(now())
    updatedAt           DateTime            @updatedAt
    components          Component[]
    quotations          Quotation[]
}

model Component {
    id                  Int                 @default(autoincrement()) @id
    label               String
    reference           String              @unique
    shortDescription    String?             @db.Text
    description         String              @db.MediumText
    price               Float
    createdAt           DateTime            @default(now())
    updatedAt           DateTime            @updatedAt
    provider            Provider            @relation(fields: [providerId], references: [id])
    unit                UsageUnit           @relation(fields: [unitId], references: [id])
    providerId          Int
    unitId              Int
    modules             Module[]
}
```

Here's the kind of data I'm getting from the front-end : 
```js
const componentsId = [1, 2, 3, 3, 3, 4];
```

And here's how I'm mapping the components to my module : 

```js
await db.module.create({
    data: {
        label,
        reference,
        shortDescription,
        description,
        components: {
            connect: componentsId.map(id => ({ id }))
        }
    }
});
```

The problem is that it will create only 4 relations, and the two other relations for the id 3 will not be created.

If anyone can help me with this, I'll be very grateful :)

**Comment by wazazaby** [comment]
So I've looked at the generated migration file for this schema, and found that it creates a unique index composed of both ids
```sql
CREATE TABLE `_ComponentToModule` (
    `A` INTEGER NOT NULL,
    `B` INTEGER NOT NULL,
UNIQUE INDEX `_ComponentToModule_AB_unique`(`A`, `B`),
INDEX `_ComponentToModule_B_index`(`B`)
) DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
``` 

I think I'll have to create an explicit model for this relation then :(

**Reply by ryands17** [reply]
@Wazazaby 👋 

Could you explain the use-case where you want to store the same id mapping multiple times?

**Reply by wazazaby** [reply]
Hi!
Yes of course. I'm building an API for the backend of a website where you can create bespoke houses. Each house is composed of multiple modules and each module is composed of multiple components (for example, a component can be 1 square meter of wooden floor). Let's say you're designing a room, and you need 10 square meters of floor, you'll at the component 10 times to the module named "Room floor".

I thought about another solution, I don't know if it's viable tho : having a counter on the relation table to set the number of times the component is needed for this module.

If you have any idea on how we could implement this differently, I'm all ears.

**Reply by ryands17** [reply]
@Wazazaby I think having a counter would be better and scalable where you could directly fetch all the components in those modules and their respective quantities. 

This is quite similar to recipes and ingredients where there could be many ingredients in a recipe and we set the quantity of those ingredients that will be added to a particular recipe.

Adding removing quantities will also be easy with [atomic operations](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#atomic-number-operations).

Let me know if this helps :)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
